File: CodingInterviews/src/main/kotlin/cn/dong/leetcode/BinarySearch.java
package cn.dong.leetcode;

import static org.junit.Assert.assertEquals;

/**
 * @author dong on 2018/11/03.
 */
public class BinarySearch {

    /**
     * @return -1 if target not exists
     */
    public static int searchFirst(int[] a, int target) {
        int low = 0;
        int high = a.length - 1;
        while (low <= high) {
            int mid = (low + high) >>> 1; // avoid overflow
            int midVal = a[mid];
            if (midVal > target) {
                high = mid - 1;
            } else if (midVal < target) {
                low = mid + 1;
            } else { // midVal == target
                if (mid == 0 || a[mid - 1] != target) {
                    return mid;
                } else {
                    high = mid - 1;
                }
            }
        }
        return -1;
    }

    public static int searchLast(int[] a, int target) {
        int low = 0;
        int high = a.length - 1;
        while (low <= high) {
            int mid = (low + high) >>> 1;
            int midVal = a[mid];
            if (midVal < target) {
                low = mid + 1;
            } else if (midVal > target) {
                high = mid - 1;
            } else { // midVal == target
                if (mid == a.length - 1 || a[mid + 1] != target) {
                    return mid;
                } else {
                    low = mid + 1;
                }
            }
        }
        return -1;
    }

    // >= target
    public static int searchFirstGreaterThanOrEqualTo(int[] a, int target) {
        int low = 0;
        int high = a.length - 1;
        while (low <= high) {
            int mid = (low + high) >>> 1;
            if (a[mid] >= target) {
                // 这是不是第一个 >= target 的元素？通过判断前一个元素确定。是则命中，否则在前面继续找
                if (mid == 0 || a[mid - 1] < target) {
                    return mid; // hit
                } else {
                    high = mid - 1;
                }
            } else { // < target
                low = mid + 1;
            }
        }
        return -1;
    }

    // <= target
    public static int searchLastLessThanOrEqualTo(int[] a, int target) {
        int low = 0;
        int high = a.length - 1;
        while (low <= high) {
            int mid = (low + high) >>> 1;
            if (a[mid] <= target) {
                // 这是不是最后一个 <= target 的元素？通过判断下一个元素确定。是则命中，否则在后面继续找
                if (mid == a.length - 1 || a[mid + 1] > target) {
                    return mid; // hit
                } else {
                    low = mid + 1;
                }
            } else { // > target
                high = mid - 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] input = {1, 2, 2, 2, 3, 4, 4, 7};
        assertEquals(0, searchFirst(input, 1));
        assertEquals(1, searchFirst(input, 2));
        assertEquals(5, searchFirst(input, 4));
        assertEquals(7, searchFirst(input, 7));
        assertEquals(-1, searchFirst(input, 0));
        assertEquals(-1, searchFirst(input, 6));

        assertEquals(0, searchLast(input, 1));
        assertEquals(3, searchLast(input, 2));
        assertEquals(6, searchLast(input, 4));
        assertEquals(7, searchLast(input, 7));
        assertEquals(-1, searchLast(input, 0));
        assertEquals(-1, searchLast(input, 6));

        assertEquals(4, searchFirstGreaterThanOrEqualTo(input, 3));
        assertEquals(7, searchFirstGreaterThanOrEqualTo(input, 5));
        assertEquals(0, searchFirstGreaterThanOrEqualTo(input, 0));
        assertEquals(-1, searchFirstGreaterThanOrEqualTo(input, 8));

        assertEquals(4, searchLastLessThanOrEqualTo(input, 3));
        assertEquals(6, searchLastLessThanOrEqualTo(input, 5));
        assertEquals(7, searchLastLessThanOrEqualTo(input, 8));
        assertEquals(0, searchLastLessThanOrEqualTo(input, 1));
        assertEquals(-1, searchLastLessThanOrEqualTo(input, 0));
    }
}

File: code/src/main/kotlin/cn/dong/geektime/BinarySearch.kt
package cn.dong.geektime

import kotlin.test.assertEquals

// [16 | 二分查找（下）：如何快速定位IP对应的省份地址？-数据结构与算法之美-极客时间](https://time.geekbang.org/column/article/42733)
/**
 * 查找第一个值等于给定值的元素的 Index
 */
fun searchFirst(array: IntArray, target: Int): Int {
    var low = 0
    var high = array.size - 1
    while (low <= high) {
        val mid = low + (high - low) / 2
        val midVal = array[mid]
        if (target == midVal) {
            if (mid > 0 && target == array[mid - 1]) {
                high = mid - 1
            } else {
                return mid
            }
        } else if (target < midVal) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return -1
}

/**
 * 查找最后一个值等于给定值的元素
 */
fun searchLast(array: IntArray, target: Int): Int {
    var low = 0
    var high = array.size - 1

    while (low <= high) {
        var mid = low + (high - low) / 2
        var midVal = array[mid]
        if (midVal == target) {
            if (mid < array.size - 1 && target == array[mid + 1]) {
                low = mid + 1
            } else {
                return mid
            }
        } else if (target < midVal) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return -1
}

/**
 * 找第一个大于等于给定值的元素
 */
fun searchFirstGreaterOrEqual(array: IntArray, target: Int): Int {
    var low = 0
    var high = array.size - 1
    while (low <= high) {
        val mid = low + (high - low) / 2
        if (array[mid] < target) {
            low = mid + 1
        } else { // array[mid] >= target
            if (mid > 0 && array[mid - 1] >= target) {
                high = mid - 1
            } else {
                return mid
            }
        }
    }
    return -1
}

/**
 * 查找最后一个小于等于给定值的元素
 */
fun searchLastLessOrEqual(array: IntArray, target: Int): Int {
    var low = 0
    var high = array.size - 1
    while (low <= high) {
        val mid = low + (high - low) / 2
        if (array[mid] > target) {
            high = mid - 1
        } else { // array[mid] <= target
            if (mid < array.size - 1 && array[mid + 1] <= target) {
                low = mid + 1
            } else {
                return mid
            }
        }
    }
    return -1
}

fun main() {
    val array = intArrayOf(1, 3, 4, 4, 4, 4, 6, 7, 8)
    assertEquals(2, searchFirst(array, 4))
    assertEquals(0, searchFirst(array, 1))
    assertEquals(5, searchLast(array, 4))

    assertEquals(6, searchFirstGreaterOrEqual(array, 5))
    assertEquals(2, searchFirstGreaterOrEqual(array, 4))
    assertEquals(1, searchFirstGreaterOrEqual(array, 2))
    assertEquals(-1, searchFirstGreaterOrEqual(array, 9))

    assertEquals(8, searchLastLessOrEqual(array, 9))
    assertEquals(-1, searchLastLessOrEqual(array, 0))
    assertEquals(5, searchLastLessOrEqual(array, 4))
    assertEquals(5, searchLastLessOrEqual(array, 5))

}